java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -426,11 +450,12 @@ public boolean exitingOnRuntimeException() {
 
   Set<LearnedConstraint> learnedConstraints(ProgramState parent) {
     ImmutableSet.Builder<LearnedConstraint> result = ImmutableSet.builder();
-    constraints.forEach((sv, childConstraint) -> {
-      Constraint parentConstraint = parent.getConstraint(sv);
-      if (parentConstraint == null || !parentConstraint.equals(childConstraint)) {
-        addLearnedConstraint(result, sv, childConstraint);
-      }
+    constraints.forEach((sv, pmap) -> {
+      pmap.forEach((domain, c) -> {
+        if (parent.getConstraint(sv, domain) != c) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)